File: implementations/csharp/protocol/tests/Specus.Protocol.Tests/PeerEgressControlVectorTests.cs
using System.Text.Json;
using Specus.Protocol.PeerEgress;

namespace Specus.Protocol.Tests;

/// <summary>
/// Replays <c>peer-egress-control-v1.json</c> against the .NET <c>egress-config</c> decoder.
/// </summary>
/// <remarks>
/// The push is produced by three server implementations and read by three clients. Its expectations
/// come from an independent reference decoder in the generator rather than from any implementation,
/// so agreeing with it is independent agreement.
/// </remarks>
public class PeerEgressControlVectorTests
{
    private static JsonDocument ReadVector()
    {
        var directory = new DirectoryInfo(AppContext.BaseDirectory);
        while (directory is not null)
        {
            var candidate = Path.Combine(
                directory.FullName, "protocol", "test-vectors", "peer-egress-control-v1.json");
            if (File.Exists(candidate))
            {
                return JsonDocument.Parse(File.ReadAllText(candidate));
            }
            directory = directory.Parent;
        }
        throw new FileNotFoundException("cannot locate peer-egress-control-v1.json");
    }

    [Fact]
    public void EgressConfigDecodeMatchesSharedVector()
    {
        using var vector = ReadVector();
        var section = vector.RootElement.GetProperty("egressConfig");
        var accept = section.GetProperty("accept");
        var reject = section.GetProperty("reject");
        Assert.True(accept.GetArrayLength() > 0, "control vector carried no accept cases");
        Assert.True(reject.GetArrayLength() > 0, "control vector carried no reject cases");

        foreach (var testCase in accept.EnumerateArray())
        {
            var name = testCase.GetProperty("name").GetString()!;
            var decoded = PeerEgressConfigMessage.Decode(testCase.GetProperty("message").GetString());
            Assert.True(decoded is not null, $"{name}: the message was refused");

            var expect = testCase.GetProperty("expect");
            Assert.True(expect.GetProperty("revision").GetInt64() == decoded!.Revision, $"{name}: revision");

            var want = expect.GetProperty("policy");
            var policy = decoded.Policy;
            Assert.True(want.GetProperty("enabled").GetBoolean() == policy.Enabled, $"{name}: enabled");
            Assert.True(want.GetProperty("scope").GetString() == policy.Scope, $"{name}: scope");
            Assert.True(
                want.GetProperty("allowedConsumerClientIds").GetArrayLength() == policy.AllowedConsumerClientIds.Count,
                $"{name}: consumers");

            var wantRules = want.GetProperty("destinationRules");
            Assert.True(wantRules.GetArrayLength() == policy.DestinationRules.Count,
                $"{name}: destination rules");
            var index = 0;
            foreach (var expectedRule in wantRules.EnumerateArray())
            {
                var rule = policy.DestinationRules[index];
                Assert.True(expectedRule.GetProperty("cidr").GetString() == rule.Cidr,
                    $"{name}: rule {index} cidr");
                Assert.True(expectedRule.GetProperty("protocols").GetArrayLength() == rule.Protocols.Count,
                    $"{name}: rule {index} protocols");
                Assert.True(expectedRule.GetProperty("portRanges").GetArrayLength() == rule.PortRanges.Count,
                    $"{name}: rule {index} port ranges");
                index++;
            }

            var limits = want.GetProperty("limits");
            Assert.True(limits.GetProperty("maxConcurrentFlows").GetInt32() == policy.Limits.MaxConcurrentFlows,
                $"{name}: maxConcurrentFlows");
            Assert.True(limits.GetProperty("maxFlowsPerConsumer").GetInt32() == policy.Limits.MaxFlowsPerConsumer,
                $"{name}: maxFlowsPerConsumer");
            Assert.True(limits.GetProperty("idleTimeoutSeconds").GetInt32() == policy.Limits.IdleTimeoutSeconds,
                $"{name}: idleTimeoutSeconds");
        }

        foreach (var testCase in reject.EnumerateArray())
        {
            var name = testCase.GetProperty("name").GetString()!;
            Assert.True(PeerEgressConfigMessage.Decode(testCase.GetProperty("message").GetString()) is null,
                $"{name}: the message was accepted");
        }
    }
}

File: implementations/java/common/src/test/java/com/theshuai/common/peeregress/PeerEgressControlVectorTests.java
package com.theshuai.common.peeregress;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import org.junit.jupiter.api.Test;

/**
 * Replays {@code peer-egress-control-v1.json} against the Java {@code egress-config} decoder.
 *
 * <p>The push is produced by three server implementations and read by three clients. Its
 * expectations come from an independent reference decoder in the generator rather than from any
 * implementation, so agreeing with it is independent agreement.
 */
class PeerEgressControlVectorTests {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static JsonNode readVector() throws IOException {
        Path directory = Path.of("").toAbsolutePath();
        for (int depth = 0; depth < 8 && directory != null; depth++) {
            Path candidate = directory.resolve("protocol").resolve("test-vectors")
                    .resolve("peer-egress-control-v1.json");
            if (Files.exists(candidate)) {
                return MAPPER.readTree(Files.readString(candidate));
            }
            directory = directory.getParent();
        }
        throw new IOException("cannot locate peer-egress-control-v1.json");
    }

    @Test
    void egressConfigDecodeMatchesSharedVector() throws IOException {
        JsonNode section = readVector().path("egressConfig");
        assertTrue(section.path("accept").size() > 0, "control vector carried no accept cases");
        assertTrue(section.path("reject").size() > 0, "control vector carried no reject cases");

        for (JsonNode testCase : section.path("accept")) {
            String name = testCase.path("name").asText();
            PeerEgressConfigMessage decoded =
                    PeerEgressConfigMessage.decode(testCase.path("message").asText());
            assertNotNull(decoded, name + ": the message was refused");

            JsonNode expect = testCase.path("expect");
            assertEquals(expect.path("revision").asLong(), decoded.revision(), name + ": revision");

            JsonNode want = expect.path("policy");
            PeerEgressPolicy policy = decoded.policy();
            assertEquals(want.path("enabled").asBoolean(), policy.isEnabled(), name + ": enabled");
            assertEquals(want.path("scope").asText(), policy.getScope(), name + ": scope");
            assertEquals(want.path("allowedConsumerClientIds").size(),
                    policy.getAllowedConsumerClientIds().size(), name + ": consumers");
            assertEquals(want.path("destinationRules").size(),
                    policy.getDestinationRules().size(), name + ": destination rules");

            for (int index = 0; index < want.path("destinationRules").size(); index++) {
                JsonNode expectedRule = want.path("destinationRules").get(index);
                PeerEgressPolicy.PeerEgressDestinationRule rule =
                        policy.getDestinationRules().get(index);
                assertEquals(expectedRule.path("cidr").asText(), rule.getCidr(),
                        name + ": rule " + index + " cidr");
                assertEquals(expectedRule.path("protocols").size(), rule.getProtocols().size(),
                        name + ": rule " + index + " protocols");
                assertEquals(expectedRule.path("portRanges").size(), rule.getPortRanges().size(),
                        name + ": rule " + index + " port ranges");
            }

            JsonNode limits = want.path("limits");
            assertEquals(limits.path("maxConcurrentFlows").asInt(),
                    policy.getLimits().getMaxConcurrentFlows(), name + ": maxConcurrentFlows");
            assertEquals(limits.path("maxFlowsPerConsumer").asInt(),
                    policy.getLimits().getMaxFlowsPerConsumer(), name + ": maxFlowsPerConsumer");
            assertEquals(limits.path("idleTimeoutSeconds").asInt(),
                    policy.getLimits().getIdleTimeoutSeconds(), name + ": idleTimeoutSeconds");
        }

        for (JsonNode testCase : section.path("reject")) {
            assertNull(PeerEgressConfigMessage.decode(testCase.path("message").asText()),
                    testCase.path("name").asText() + ": the message was accepted");
        }
    }
}
